### ISSUE TITLE
Implement Sequential Quasi Monte Carlo

### ISSUE THREAD

**Issue by AdrienCorenflos** [state-space-models/cuthbert#255]
Implement the method listed in https://arxiv.org/abs/1402.4039

This requires a particular structure slightly differing from the current state of SMC: resampling and propagation are done via consumption of low-discrepancy sequences.

**Comment by ryantjx** [comment]
@AdrienCorenflos Hi Adrien, I would like to take on this task!

**Comment by ryantjx** [comment]
hi @AdrienCorenflos I am inclined to create sqmc as a standalone filter under `cuthbert.smc.sqmc`. sqmc resampling is coupled with the propagation so its not possible to be implemented with the same design in `filter_combine`, which has resampling followed by propagation. The sqmc's `filter_combine` will run through the SQMC step of:

1. generate N RQMC points of 1+d dimension
2. Hilbert sort previous N particles to get Hilbert-ordered weights
3. Sort RQMC by the first coordinate
4. select ancestors by inverse CDF using first coordinates and Hilbert-ordered weights
5. Propagate ancestors through a deterministic function `propagate_transform` (instead of `propagate_sample`, which is stochastic) using d coordinates
6. Reweight observation
7. update log-normalization constant

at the end it should return a `Filter` object that can be built with `cuthbert.filtering.filter`

What are your thoughts on this?

**Comment by AdrienCorenflos** [comment]
I think it's likely going to be the outcome yes, but it's a bit rushed. The general structure should be to refactor the resamplings so they become compatible (i.e. introduce an intermediate version that takes uniforms), then deal with sorting, then only propagation.

**Comment by ryantjx** [comment]
Understood, I will write a copy of this in a fork/branch in the meantime. I think the refactor should happen [multinomial.py](https://github.com/state-space-models/cuthbert/blob/main/cuthbertlib/resampling/multinomial.py) because it uses the inverse_cdf to resample too. So it would look something like this

```
@partial(resampling_decorator, name="Multinomial", desc=_DESCRIPTION)
def resampling(
    key: Array, logits: ArrayLike, positions: ArrayTreeLike, n: int
) -> tuple[Array, Array, ArrayTree]:
    key_uniforms, key_shuffle = random.split(key)
    sorted_uniforms = _sorted_uniforms(key_uniforms, n)
    idx, logits_out = resample_from_uniform(sorted_uniforms, logits, positions)
    idx = random.permutation(key_shuffle, idx)
    return idx, logits_out, apply_resampling_indices(positions, idx)

def resample_from_uniform(sorted_uniforms, logits, position):
    idx = inverse_cdf(sorted_uniforms, logits)
    logits_out = jnp.zeros_like(sorted_uniforms)
    return idx, logits_out
```

where `resample_from_uniform` can then be used in SQMC. Is this on the right track you were thinking of?
